# Исследование надежности заемщиков


**Цель исследования** - нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

**Описание данных**:
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

**План работы**:
* изучить общую информацию о данных;
* предобработка данных (удаление пропусков, обработка аномальных значений, изменение типов данных, обработка дубликатов, категоризация данных);
* исследование данных.

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
data_pivot_children = data.pivot_table(index=['children'], values='debt', aggfunc={'sum', 'count', 'mean'}) # создание сводной таблицы
data_pivot_children

count      mean     sum
children                           
0         14091.0  0.075438  1063.0
1          4808.0  0.092346   444.0
2          2052.0  0.094542   194.0
3           330.0  0.081818    27.0
4            41.0  0.097561     4.0
5             9.0  0.000000     0.0

**Вывод:** По результатам сводной таблицы видно, что среди всех клиентов процент должников не сильно разбросан - 8%-10%.  Клиентов с 5 детьми не учитываю, их слишком мало. Клиетов без детей по количеству больше всех - 14 091, должников из них 1063, 7,5%. Клиентов с одним ребенком (почти в три раза меньше, чем клиентов без детей) - 4808, из них долгов по возврату кредита в срок - 444, 9,2% по процентам - это больше, чем у клиентов без детей. С двумя детьми еще меньше клиентов, всего 2052 из них долги имеют 9,5% клиентов - 194 клиентов. С 3 детьми должников среди клиентов 8,1%. По проценту должников, самый плохой показатель у клиентов с 4 детьми - 9,8%, однако таких клиентов очень мало, в сравнении с другими.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
data_pivot_family_status = data.pivot_table(index=['family_status'], values='debt', aggfunc={'sum', 'count', 'mean'}) # создание сводной таблицы
data_pivot_family_status

count      mean    sum
family_status                                  
Не женат / не замужем   2796.0  0.097639  273.0
в разводе               1189.0  0.070648   84.0
вдовец / вдова           951.0  0.066246   63.0
гражданский брак        4134.0  0.093130  385.0
женат / замужем        12261.0  0.075606  927.0

**Вывод:** согласно таблице, самый высокий процент невозврата кредита в срок у клиентов с семейным статусом не женат/ не замужем 9,8% из 2796 клиентов у 273 долги. Чуть лучше кредит возвращают клиенты в гражданском браке, процент должников - 9,3, но таких клиентов больше (4134 клиента и долгов соответственно тоже больше - 385). По количеству клиентов и долгов,больше всех клиентов женатых и замужних, их 12261, а долгов 927, это 7,6%. Самый низкий процент должников у клиентов в разводе их 1189, должников 84, это 7,1%, у клиентов в статусе вдовец/вдова 6,6% должников, 63 должника из 951 клиента.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [25]:
data_pivot_total_income = data.groupby('total_income_category').agg({'debt': ['count', 'mean', 'sum']}) # группирую данные по столбцу 'debt'
data_pivot_total_income

debt                
                       count      mean   sum
total_income_category                       
A                         25  0.080000     2
B                       5014  0.070602   354
C                      15921  0.084982  1353
D                        349  0.060172    21
E                         22  0.090909     2

**Вывод:** Расшифровка к уровню дохода:
* 1000001 и выше — 'A'
* 200001–1000000 — 'B';
* 50001–200000 — 'C';
* 30001–50000 — 'D';
* 0–30000 — 'E';

По таблице видно, что больше всего процент должников среди клиентов с доходом до 30 000 руб, категория "Е", 9,1%, но их совсем немного. Чуть меньше процент должников у клиентов с доходом от 50 001 до 200 000 рублей (категория "С"), их 8,5%. 
8% должников среди клиентов с доходом категории "А" свыше 1млн рублей таких клиентов тоже немного. 
Высокая возвратность по кредиту у клиентов с доходом от 30 001 до 50 000 рублей, всего 6% должников, чуть выше процент должников у клиентов с доходом от 200 001 до 1млн рублей - 7%.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
data_pivot_purpose = data.groupby('purpose_category').agg({'debt': ['count', 'mean', 'sum']}) # группирую данные
data_pivot_purpose


debt               
                          count      mean  sum
purpose_category                              
операции с автомобилем     4279  0.093480  400
операции с недвижимостью  10751  0.072551  780
получение образования      3988  0.092528  369
проведение свадьбы         2313  0.079118  183

**Вывод:** Больше всего просрочек по кредитам на операции с автомобилем, 9,4% должников, на получение образования должников чуть меньше, 9,3%. По кредитам с целью проведение свадьбы нарушают срок оплаты 8% клиентов, самый низкий процент должников у кредитов, взяты на операции с недвижимостью - 7,3%.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски в исходных данных могут быть потому что пользователь не указал о себе сведения или система сбора информации дала сбой, еще пропуски могут оставлять умышленно, рассчитывая на автозаполнение.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана делит набор данных на две равные половины, она устойчива к аномально большим и малым значениям. Медианное значение лучше всего отражает выборку в целом, если есть пропуски в данных. Среднее арифметическое отражает ситуацию некорректно, так как не учитывает разброс данных.

### Шаг 4: общий вывод.

В ходе проекта, после предобработки данных, клиенты были разделены на категории по доходам, также все цели кредита были собраны и разделены на 4 основные категории.

Если сравнить процент должников по количеству детей: без детей - 7,5% должников, с одним ребенком - 9,2%, с двумя детьми - 9,5%, с тремя детьми - 8,1%, четыре ребенка - 9,8%. Процент растем, с тремя детьми немного ниже, возможно связано с государственной поддержкой многодетных семей. В основном получается, что чем больше детей у клиентов, тем больше процент долгов по возврату кредита в срок.

Посмотрим по семейному статусу процент должников: 9,8% клиентов-должников не женаты/не замужем, 9,3% - в гражданском браке, 7,6% - женат/замужем, 7,1% - в разводе, 6,6% - вдовец/вдова. Получается, что клиенты до официального вступления в брак к возврату кредита в срок относятся менее ответственно, а если вдруг клиент в разводе или овдовел, то все-таки чаще возвращает кредит в срок.

Посмотрим процент долгов в зависимости от дохода клиента: 9,1% должников у клиентов с доходом до 30 000 руб., 8,5% у клиентов с доходом - 50 001-200 000 руб, 8% должников с доходом свыше 1млн.руб., 7% должников среди клиентов с доходом 200 001 - 1млн руб, 6% должников с доходом 30 001 - 50 000 руб. Клиентов с доходом до 30 000 руб и свыше 1млн очень мало, до 25 человек, поэтому учитывать их не буду. Более ответственные клиенты с доходом от 30 001 до 50 000 рублей. Менее ответственные - клиенты с доходом 50 001 - 200 000 руб.

Посмотри как цели кредита влияют на возврат кредита в срок. На операции с автомобилем - 9,4% должников, на получение образования - 9,3%, на проведение свадьбы - 8% должников, на операции с недвижимостью - 7,3%. Получается, что менее ответственно клиенты относятся к возврату кредитов связанных с операциями с автомобилями и с получением образования. С кредитом на недвижимость клиенты чаще не нарушают сроки по возврату.

По итогам проведения проекта, можно сделать вывод, что больше всего долгов по возврату кредита в срок у клиентов с детьми, ввиду дополнительной финансовой нагрузки, в семейном статусе не женат/не замужем, предположу, что неполная семья, из-за чего финансовая нагрузка на одного родитеся еще больше возрастает, со средним доходом от 50 001 до 200 000 рублей, на потребительский кредит (автомобиль, образование, свадьба).
